# Regras de Testes de Objetos de Dom√≠nio

O dom√≠nio √© o cora√ß√£o do StarDust. Como ele encapsula as regras de neg√≥cio essenciais, garantimos que essa camada possua uma alta cobertura de testes e siga padr√µes rigorosos de qualidade.

Este documento detalha como testar **Entidades**, **Estruturas (Value Objects)** e **Agregados**.

## üìç Localiza√ß√£o e Nomenclatura

Os arquivos de teste devem ser colocados em uma subpasta `tests` dentro do diret√≥rio do objeto correspondente.

- **Entidades**: `packages/core/src/[modulo]/domain/entities/tests/[NomeEntidade].test.ts`
- **Estruturas**: `packages/core/src/[modulo]/domain/structures/tests/[NomeEstrutura].test.ts`

O sufixo deve ser sempre `.test.ts`.

## ü•∏ O Papel dos Fakers

Para agilizar a cria√ß√£o de testes e evitar o acoplamento com a estrutura interna de DTOs, utilizamos **Fakers**. Cada objeto de dom√≠nio complexo deve ter seu Faker correspondente.

- **Uso**: Sempre prefira `EntidadeFaker.fake()` em vez de instanciar manualmente no teste.
- **Sobrescrita**: Utilize o par√¢metro de overrides para definir apenas os dados relevantes para o cen√°rio do teste.

```ts
// Bom: Foca apenas no que importa para o teste
const user = UsersFaker.fake({ coins: 0 });

// Ruim: Verboso e propenso a quebras por mudan√ßas no DTO
const user = User.create({
  id: 'any-id',
  name: 'Any Name',
  email: 'any@email.com',
  coins: 0,
  // ... dezenas de outros campos
});
```

---

## üèóÔ∏è Testando Entidades (Entities)

Entidades s√£o mut√°veis e identificadas por um `id`. Os testes devem focar no **comportamento** e nas **mudan√ßas de estado**.

### O que testar:
1.  **M√©todos de Neg√≥cio**: Verifique se o estado interno (`props`) muda corretamente ap√≥s a execu√ß√£o de um m√©todo.
2.  **Regras de Valida√ß√£o**: Verifique se exce√ß√µes (Domain Errors) s√£o lan√ßadas quando pr√©-condi√ß√µes n√£o s√£o atendidas.
3.  **Garantia de Identidade**: Verifique se a identidade √© preservada.
4.  **Convers√£o para DTO**: Garanta que o m√©todo `dto` ou getter `dto` retorna a representa√ß√£o correta.

### Exemplo:
```ts
describe('User Entity', () => {
  it('should increase coins when user earns coins', () => {
    const user = UsersFaker.fake({ coins: 100 });
    
    user.earnCoins(Integer.create(50));
    
    expect(user.coins.value).toBe(150);
  });

  it('should throw NotEnoughCoinsError when losing more coins than available', () => {
    const user = UsersFaker.fake({ coins: 10 });
    
    expect(() => user.loseCoins(Integer.create(20)))
      .toThrow(NotEnoughCoinsError);
  });
});
```

---

## üíé Testando Estruturas (Structures)

Estruturas (Value Objects) s√£o imut√°veis. O foco dos testes deve ser a **valida√ß√£o na cria√ß√£o** e a **corretude das opera√ß√µes**.

### O que testar:
1.  **Cria√ß√£o (Factory Method)**: Teste o m√©todo `static create()` com valores v√°lidos e inv√°lidos.
2.  **Imutabilidade**: Garanta que m√©todos que "alteram" o valor retornem uma **nova inst√¢ncia**, sem modificar a original.
3.  **Compara√ß√£o**: Teste m√©todos de igualdade como `isEqualTo`.
4.  **L√≥gica Interna**: Teste algoritmos encapsulados (ex: c√°lculo de XP, formata√ß√£o).

### Exemplo:
```ts
describe('Email Structure', () => {
  it('should be created from a valid email string', () => {
    expect(() => Email.create('valid@email.com')).not.toThrow();
  });

  it('should throw ValidationError for invalid email formats', () => {
    expect(() => Email.create('invalid-email')).toThrow(ValidationError);
  });
});

describe('Level Structure', () => {
  it('should return a new instance when leveling up', () => {
    const level = Level.create(1);
    const nextLevel = level.up(currentXp, earnedXp);
    
    expect(nextLevel).not.toBe(level); // Imutabilidade
    expect(nextLevel.value).toBe(2);
  });
});
```

---

## üß© Testando Agregados (Aggregates)

Como agregados apenas encapsulam uma rela√ß√£o entre entidades, seus testes geralmente s√£o simples e focam em garantir que a entidade agregada est√° sendo carregada ou referenciada corretamente.

---

## üöÄ Boas Pr√°ticas

1.  **Isolamento**: Testes de dom√≠nio **nunca** devem depender de infraestrutura (banco de dados, APIs).
2.  **Tipagem Forte**: Utilize os tipos de dom√≠nio (como `Integer`, `Id`, `Logical`) nas asser√ß√µes sempre que poss√≠vel.
3.  **Arrange-Act-Assert**:
    - **Arrange**: Prepare o estado usando Fakers.
    - **Act**: Chame o m√©todo de dom√≠nio.
    - **Assert**: Verifique o resultado ou a mudan√ßa de estado.
4.  **Use Spies com modera√ß√£o**: Se um objeto de dom√≠nio interage com outro objeto complexo passado por par√¢metro, use `jest.spyOn` para verificar a intera√ß√£o se necess√°rio.

## Tooling

- Runner de testes: Jest (principalmente no pacote `@stardust/core`).
- Geracao de dados: fakers do core + `@faker-js/faker` (quando aplicavel).
- Comandos:
  - `npm run test -w @stardust/core`
  - `npm run typecheck -w @stardust/core`
- Referencia geral: `documentation/tooling.md`.
